File: Binary Search/50. Pow(x, n).py
"""
1. Clarification
2. Possible solutions
 - library function
 - built-in calculation
 - brute force
 - fast exp, recursive (divide and conquer)
 - fast exp, iterative
3. Coding
4. Tests
"""

# # T=O(1)
# class Solution:
#     def myPow(self, x: float, n: int) -> float:
#         return pow(x, n)

# # T=O(lgn)
# class Solution:
#     def myPow(self, x: float, n: int) -> float:
#         return x**n

# # T=O(n), Time Limit Exceeded
# class Solution:
#     def myPow(self, x: float, n: int) -> float:
#         ans = 1.0
#         flag = True
#         if n < 0:
#             flag = False
#             n = -n
#         for _ in range(n):
#             ans *= x
#         return 1/ans if flag == False else ans

# # T=O(lgn), S=O(lgn)
# class Solution:
#     def myPow(self, x: float, n: int) -> float:
#         if not n: return 1
#         if n < 0: return 1 / self.myPow(x, -n)
#         if n % 2: return x * self.myPow(x, n-1)
#         return self.myPow(x*x, n//2)

# T=O(lgn), S=O(1)
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            x = 1 / x
            n = -n
        pow = 1
        while n:
            if n & 1:
                pow *= x
            x *= x
            n >>= 1
        return pow

File: 面试经典 150 题/数学/50. Pow(x, n).cpp
typedef long long LL;

class Solution {
  double qmi(double a, LL b) {
    if (b == 0) return 1.0;

    double res = 1.0;
    while (b) {
      if (b & 1) res = res * a;
      a = a * a;
      b >>= 1;
    }
    return res;
  }

public:
  double myPow(double x, int n) {
    LL b = n;
    if (b > 0) return qmi(x, b);
    else return 1.0 / qmi(x, -b);
  }
};